receiver/prometheusremotewritereceiver/receiver.go

```diff
@@ -388,23 +392,115 @@ func addNumberDatapoints(datapoints pmetric.NumberDataPointSlice, ls labels.Labe
-	// TODO: Implement this function
+func addExponentialHistogramDatapoints(datapoints pmetric.ExponentialHistogramDataPointSlice, ls labels.Labels, ts writev2.TimeSeries) {
+	for _, histogram := range ts.Histograms {
+		dp := datapoints.AppendEmpty()
+		dp.SetStartTimestamp(pcommon.Timestamp(ts.CreatedTimestamp * int64(time.Millisecond)))
+		dp.SetTimestamp(pcommon.Timestamp(histogram.Timestamp * int64(time.Millisecond)))
+
+		// The difference between float and integer histograms is that float histograms are stored as absolute counts
+		// while integer histograms are stored as deltas.
+		if histogram.IsFloatHistogram() {
+			// Float histograms
+			if len(histogram.PositiveSpans) > 0 {
+				dp.Positive().SetOffset(histogram.PositiveSpans[0].Offset - 1) // -1 because OTEL offset are for the lower bound, not the upper bound
+				convertAbsoluteBuckets(histogram.PositiveSpans, histogram.PositiveCounts, dp.Positive().BucketCounts())
+			}
+			if len(histogram.NegativeSpans) > 0 {
+				dp.Negative().SetOffset(histogram.NegativeSpans[0].Offset - 1) // -1 because OTEL offset are for the lower bound, not the upper bound
+				convertAbsoluteBuckets(histogram.NegativeSpans, histogram.NegativeCounts, dp.Negative().BucketCounts())
+			}
+
+			dp.SetScale(histogram.Schema)
+			dp.SetZeroThreshold(histogram.ZeroThreshold)
+			zeroCountFloat := histogram.GetZeroCountFloat()
+			dp.SetZeroCount(uint64(zeroCountFloat))
+			dp.SetSum(histogram.Sum)
+			countFloat := histogram.GetCountFloat()
+			dp.SetCount(uint64(countFloat))
+		} else {
+			// Integer histograms
+			if len(histogram.PositiveSpans) > 0 {
+				dp.Positive().SetOffset(histogram.PositiveSpans[0].Offset - 1) // -1 because OTEL offset are for the lower bound, not the upper bound
+				convertDeltaBuckets(histogram.PositiveSpans, histogram.PositiveDeltas, dp.Positive().BucketCounts())
+			}
+			if len(histogram.NegativeSpans) > 0 {
+				dp.Negative().SetOffset(histogram.NegativeSpans[0].Offset - 1) // -1 because OTEL offset are for the lower bound, not the upper bound
+				convertDeltaBuckets(histogram.NegativeSpans, histogram.NegativeDeltas, dp.Negative().BucketCounts())
+			}
+
+			dp.SetScale(histogram.Schema)
+			dp.SetZeroThreshold(histogram.ZeroThreshold)
+			zeroCountInt := histogram.GetZeroCountInt()
+			dp.SetZeroCount(zeroCountInt)
+			dp.SetSum(histogram.Sum)
+			countInt := histogram.GetCountInt()
+			dp.SetCount(countInt)
+		}
+
+		attributes := dp.Attributes()
+		extractAttributes(ls).CopyTo(attributes)
+	}
+}
+
+// convertDeltaBuckets converts Prometheus native histogram spans and deltas to OpenTelemetry bucket counts
+// For integer buckets, the values are deltas between the buckets. i.e a bucket list of 1,2,-2 would correspond to a bucket count of 1,3,1
+func convertDeltaBuckets(spans []writev2.BucketSpan, deltas []int64, buckets pcommon.UInt64Slice) {
+	buckets.EnsureCapacity(len(deltas))
```

Comment:
`len(deltas)` is the lower bound , to get the exact number of buckets, we should iterate the spans: add up all span lengths + the size of gaps (offsets).

Reply:
[94143ab](https://github.com/open-telemetry/opentelemetry-collector-contrib/pull/39864/commits/94143aba4361a2c01d0130d70297e370d3bce395)